message/infrastructure/sql/subscriber.go

```diff
@@ -0,0 +1,326 @@
+package sql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrSubscriberClosed = errors.New("subscriber is closed")
+)
+
+type SubscriberConfig struct {
+	Logger        watermill.LoggerAdapter
+	ConsumerGroup string
+
+	// PollInterval is the interval between subsequent SELECT queries. Must be non-negative. Defaults to 1s.
+	PollInterval time.Duration
+
+	// ResendInterval is the time to wait before resending a nacked message. Must be non-negative. Defaults to 1s.
+	ResendInterval time.Duration
+
+	// MessagesTable is the name of the table that stores Watermill messages as rows. Defaults to `messages`.
+	MessagesTable string
+	// MessageOffsetsTable is the name of the table that stores the offsets of messages read by each consumer group.
+	// Defaults to `offsets_acked`.
+	MessageOffsetsTable string
+
+	// SchemaAdapter provides the schema-dependent queries and arguments for them, based on topic/message etc.
+	SchemaAdapter SchemaAdapter
+}
+
+func (c *SubscriberConfig) setDefaults() {
+	if c.Logger == nil {
+		c.Logger = watermill.NopLogger{}
+	}
+	if c.PollInterval == 0 {
+		c.PollInterval = time.Second
+	}
+	if c.ResendInterval == 0 {
+		c.ResendInterval = time.Second
+	}
+	if c.MessagesTable == "" {
```

Comment:
isn't it duplicated with SchemaAdapter?

Reply:
removed from both subscriber and publisher